### ISSUE TITLE
Set the default build configuration and RuntimeConfiguration to Release

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#33240]
Working in libraries, I now always have to add the /p:RuntimeConfiguration=Release property to use the Release version of either coreclr/mono. Can we default it to Release (which also means it builds by default on Release)?

**EDIT: The updated ask is to change the default build configuration from Debug to Release and the default RuntimeConfiguration switch to Release.**

cc @jkotas @stephentoub @marek-safar

**Comment by jkotas** [comment]
Well, I have opposite problem when working on coreclr. I think we should default *everything* to release, and then everybody can build the part that they care about as debug.

**Comment by SamMonoRT** [comment]
I like this proposal of building everything in release by default. @akoeplinger - any opinions ?

**Comment by ViktorHofer** [comment]
looping in some other folks: @danmosemsft @ericstj @dotnet/runtime-infrastructure @filipnavara

**Comment by ViktorHofer** [comment]
cc @dotnet/ncl

**Comment by akoeplinger** [comment]
Wasn't the original problem that Debug builds of CoreCLR are extremely slow (which isn't the case on Mono) so can't really be used for libraries dev work.

If I remember correctly there was some discussion about removing Debug from CoreCLR and making Checked the "new" Debug?

**Comment by ViktorHofer** [comment]
> If I remember correctly there was some discussion about removing Debug from CoreCLR and making Checked the "new" Debug?

https://github.com/dotnet/runtime/issues/31852

**Comment by jkotas** [comment]
CoreCLR checked is still noticeably slower than release for ordinary development under libraries.

**Comment by stephentoub** [comment]
I'd personally also like it if the default for everything was release.

**Comment by ViktorHofer** [comment]
Updated the title to reflect the updated ask.

**Comment by ericstj** [comment]
I'm also fine with this.  A side thought: should we make it easier to switch between release/debug during testing by using a single testhost folder?  Part of me cringes at the thought, but if its intentional and doesn't try to be incremental it might be better behavior.

**Comment by marek-safar** [comment]
>we should default everything to release

Yes, please!

**Comment by eerhardt** [comment]
If the default is Release, how will that affect the local debugging story? I very often need to debug my code, and when I use Release builds, the debugging experience has always been degraded.

I don't believe changing the default to Release is going to solve the underlying issue of wanting to use a Debug library from `src\libraries`, and a Release runtime from `src\coreclr`. Especially for developers like me who like to use Visual Studio to build and test in Test Explorer.

**Comment by ericstj** [comment]
@eerhardt that's exactly the scenario I was talking about with https://github.com/dotnet/runtime/issues/33240#issuecomment-595520913.  Do you typically want all IL to be debug or just the IL built by the SLN you are debugging?

**Comment by eerhardt** [comment]
Of course it always depends how far I need to debug something. If my library depends on another library, and I need to debug it, I probably want a Debug version of the referenced library. I guess "normally" you probably only need to debug the library you are working with, until you don't.

I guess locally, I want as much built Debug as possible, until the perf slows down enough that makes the build unusable. Previously that was all of corefx was Debug, and all of coreclr was Release, which seemed to work fine.

**Comment by stephentoub** [comment]
The default being release shouldn't affect this scenario any more than the default being debug. Today with everything defaulting to debug, you need to override the runtime to build as release. Tomorrow with everything defaulting to release, you need to override the libraries to build as debug. Seems no worse for that scenario to me.

**Comment by eerhardt** [comment]
> The default being release shouldn't affect this scenario any more than the default being debug.

You're 100% correct. Which is why I said above that changing the default of `Configuration` isn't fixing the underlying issue here.

The issue is that `RuntimeConfiguration` gets defaulted to `Configuration` when it isn't explicitly set. 

https://github.com/dotnet/runtime/blob/fef9ca70f6e7268dfc9e0f600433f5293932f166/eng/Configurations.props#L4

When you are in VS, you don't have the ability to set `RuntimeConfiguration`, but you can set `Configuration`.

If you just change the default of `Configuration` to `Release`, and leave `RuntimeConfiguration` defaulting to whatever `Configuration` is, now when I build in VS for `Debug`, it will try to use a `Debug` coreclr.

**Comment by ericstj** [comment]
> now when I build in VS for Debug, it will try to use a Debug coreclr.

Not exactly.  The coreclr is not decided by the library project builds.  It's decided by the runtime.depproj which builds out the test shared framework.  If we remove `Configuration` from the path of the test shared framework then you could build things as Debug from VS and it will run on whatever pre-existing testshared framework you already have.

**Comment by eerhardt** [comment]
At the end of the day, I would like an easier way to use one or more Debug libraries with a Release runtime. If we can make the dev experience (especially the VS experience) better, I’m all for it.

**Comment by ericstj** [comment]
Ok, so let's set requirements.

1. Default build of runtime uses `release` for coreclr and libraries.
2. Building a single library as debug permits testing that library as debug without rebuilding the entire repo.

Sound good?

**Comment by stephentoub** [comment]
> Sound good?

👍

**Comment by eerhardt** [comment]
I'm good with those requirements.

> If we remove Configuration from the path of the test shared framework

Will we also need to make a change so `<ReferenceFromRuntime Include="System.Private.CoreLib" />` will work without having to build CoreLib for Debug?

**Comment by ViktorHofer** [comment]
> Will we also need to make a change so <ReferenceFromRuntime Include="System.Private.CoreLib" /> will work without having to build CoreLib for Debug?

`ReferenceFromRuntime` uses `RuntimeConfiguration` to resolve the runtime path.

**Comment by ericstj** [comment]
I'm ok relaxing both of these so long as we never copy the bits around and they are used only for reference.  It's ok to fallback to allowing cross-configuration references (IIRC we even omit configuration from RefPath)

**Comment by halter73** [comment]
If we make this change, would we do anything about the Debug.Assert's in the runtime? I know most (maybe all?) of our automated tests use release builds already, but at least Debug.Asserts fail during local development with the default build configuration today.

**Comment by jkotas** [comment]
> I know most (maybe all?) of our automated tests use release builds already

dotnet/runtime does a ton of automated testing with debug/checked builds.

**Comment by trylek** [comment]
@jkotas is right that multiple runtime pipelines - in fact most of the common pipelines - use checked builds with assertions enabled. I however think that @halter73 has a valid point that fully switching over to release incurs a bit of local testing degradation in terms of silently hiding assertion failures. I'm not sure how impactful that is - I can easily imagine that new users making simple changes should be getting more or less the same results in both build configurations and more advanced users can easily figure out how to control their local build configuration with finer granularity.
